Delay opcode fetching for packet helpers until receiving network data

valarnin opened the pull request:
@ngld I think this solves the issues with opcodes loading before being ready, as we were discussing in Discord.

Needs more testing, I basically only managed to test `ActorCastExtra` before Windows decided that 3 weeks of uptime was enough and started dying on me.

Basically this intrinsically ties the loading of opcodes to when a packet is received, so `FFXIV_ACT_Plugin` is always in a good/ready state when loading.

It doesn't solve users having ACT running before a game update, and then still having the same instance running after the update, but then opcodes would have changed and require an update/ACT restart anyways in that case.

wexxlee reviewed (commented):
A few things I noticed while testing this:

1. I was able to see `ActorCastExtra` (0x107) lines being written out correctly, but no `AbilityExtra` (0x108) lines were written for the paired 0x15 lines.  I didn't go exhaustively test the other Machina-based custom log lines (and the non-Machina lines like `Countdown` and `CountdownCancel` seemed to work fine.)

2. I tried to reproduce test conditions that would involve an opcode update without a version bump (e.g. patch day).  I back-dated the global game version in both `/resoures/opcodes.jsonc` and in `RainbowMage.OverlayPlugin.config.json`, and with the current game client running, I launched ACT.  The following error messages were written to the OP log:
```
[7/26/2024 9:02:02 PM] Error: No config opcodes for game region Global, version 2024.07.10.0001.0000
[7/26/2024 9:02:02 PM] Error: No config opcodes for game region Global, version 2024.07.10.0001.0000
[7/26/2024 9:02:02 PM] Error: No config opcodes for game region Global, version 2024.07.10.0001.0000
```

(There may have been more, but we seem to impose a max exception count of 3 for the OP log.)

This seemed to be largely cosmetic, since closing ACT resulted in the (proper/current/correct) opcodes being downloaded and written to `RainbowMage.OverlayPlugin.config.json`, and when ACT is reopened, the errors are not repeated and OP works correctly (-ish -- see no. 1 above though).

However, the `/resources/opcodes.jsonc` file remains outdated with my backdate change.  (I can't recall if that's intentional and we just never update that file once the plugin version is downloaded, or if that's a missed file update somewhere.)

I probably need to do more testing on this, but wanted to note findings so far.

valarnin commented:
> 1. I was able to see `ActorCastExtra` (0x107) lines being written out correctly, but no `AbilityExtra` (0x108) lines were written for the paired 0x15 lines.  I didn't go exhaustively test the other Machina-based custom log lines (and the non-Machina lines like `Countdown` and `CountdownCancel` seemed to work fine.)

Fixed with commit a9d050e.



> 2. I tried to reproduce test conditions that would involve an opcode update without a version bump (e.g. patch day).  I back-dated the global game version in both `/resoures/opcodes.jsonc` and in `RainbowMage.OverlayPlugin.config.json`, and with the current game client running, I launched ACT.  The following error messages were written to the OP log:

This is sort-of intended behavior, and the "error" message should be adjusted to "info" level instead and updated to something like `No config opcodes for game region Global, version 2024.07.10.0001.0000. Checking for opcodes online next.`, or something along those lines. Feel free to suggest a better message here.

It should only be an "error"-level message if all attempts at fetching opcodes fail.

wexxlee commented:
> This is sort-of intended behavior, and the "error" message should be adjusted to "info" level instead and updated to something like `No config opcodes for game region Global, version 2024.07.10.0001.0000. Checking for opcodes online next.`, or something along those lines. Feel free to suggest a better message here.
> 
> It should only be an "error"-level message if all attempts at fetching opcodes fail.

Sure, that seems like a good approach & text (maybe 'No **cached** opcodes...`, just so it's extra clear where the missing data was.  Should there also be a logic adjustment somewhere so the same error (now info) message isn't output multiple times?

valarnin commented:
> Sure, that seems like a good approach & text (maybe 'No **cached** opcodes...`, just so it's extra clear where the missing data was. Should there also be a logic adjustment somewhere so the same error (now info) message isn't output multiple times?

The actual logic is in `OverlayPluginLogLines.cs`:

```cs
        private IOpcodeConfigEntry GetOpcode(string name, Opcodes opcodes, string version, string opcodeType, MachinaRegion machinaRegion)
        {
            if (opcodes == null)
                return null;

            if (opcodes.TryGetValue(machinaRegion, out var regionOpcodes))
            {
                if (regionOpcodes.TryGetValue(version, out var versionOpcodes))
                {
                    if (versionOpcodes.TryGetValue(name, out var opcode))
                    {
                        return opcode;
                    }
                    else
                    {
                        LogException($"No {opcodeType} opcode for game region {machinaRegion}, version {version}, opcode name {name}");
                    }
                }
                else
                {
                    LogException($"No {opcodeType} opcodes for game region {machinaRegion}, version {version}");
                }
            }
            else
            {
                LogException($"No {opcodeType} opcodes for game region {machinaRegion}");
            }

            return null;
        }
```

The outer two logged messages should probably be `Info` level instead of `Error` level. Meaning that "nothing for this region" should be info, "nothing for this region and version" should be info, and "nothing for this region, version, and opcode name" should be an error (or maybe just warning?).

The `LogException(string message)` method should probably be updated to `Log(LogLevel level, string message)` instead, and count of logged messages should be done per level instead of globally.

Finally, later in the same file during this logic:

```cs
                    // Try once to get this remotely, but only if this opcode or version is missing.
                    // TODO: we could consider getting this once always too, but for now
                    // if we ever have an incorrect (but present) opcode, another release is required.
                    if (opcode == null && !haveAttemptedOpcodeDownload)
                    {
                        haveAttemptedOpcodeDownload = true;
                        SaveRemoteOpcodesToConfig();
                        return GetOpcode(name, opcodesConfig, version, "config", machinaRegion);
                    }
```

If the last `GetOpcode` call doesn't return an opcode entry, log an actual error.

valarnin commented:
I went with a slightly different approach, basically just "only log each unique error once". There's a slight memory overhead with this approach (maybe a couple kb?) but I think this is better for helping users if they have an issue, rather than only logging the first three errors from the class.
